SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,435 @@
+/// ```
+/// %base1 = ref_element_addr %ref
+/// %base2 = alloc_stack $S
+/// %base3 = global_addr @gaddr
+/// %addr1 = struct_element_addr %base1
+/// %access1 = store %v1 to [trivial] %addr1   // accessed address is offset from base
+/// %access2 = store %v2 to [trivial] %base2   // accessed address is base itself
+/// ```
+/// The base address is never inside an access scope.
+struct AccessBase : CustomStringConvertible {
+  let baseAddress: Value
+  let storageKind: AccessStorageKind
+  
+  var description: String {
+    "\(storageKind) - \(baseAddress)"
+  }
+  
+  var referenceWithPathComponent: (ref: Value, comp: (SmallProjectionPath.FieldKind, Int))? {
+    switch storageKind {
+    case .Box, .Tail, .Class:
+      switch baseAddress {
+      case let rea as RefElementAddrInst:
+        return (rea.operand, (.classField, rea.fieldIndex))
+      case let pb as ProjectBoxInst:
+        return (pb.operand, (.classField, pb.fieldIndex))
+      case let rta as RefTailAddrInst:
+        return (rta.operand,(.tailElements, 0))
+      default:
+        fatalError("Inconsistent state")
+      }
+    default:
+      return nil
+    }
+  }
+  
+  var reference: Value? { referenceWithPathComponent?.ref }
+  
+  var isLet: Bool {
+    switch baseAddress {
+    case is RefElementAddrInst:
+      fatalError() // FIXME: add support for looking up types in libswift etc
+    default:
+      fatalError()
+    }
+  }
+  
+  var isLocal: Bool {
+    switch storageKind {
+    case .Box:
+      return baseAddress is AllocBoxInst
+    case .Stack:
+      return true
+    default:
+      return false
+    }
+  }
+  
+  /// Returns `true` if we can reliably compare for equality this `AccessBase`
+  /// with another `AccessBase`
+  var isUniquelyIdentified: Bool {
+    switch storageKind {
+    case .Box:
+      return baseAddress is AllocBoxInst
+    case .Stack, .Global:
+      return true
+    case .Argument:
+      return (baseAddress as! FunctionArgument).isExclusiveIndirectParameter
+    default:
+      return false
+    }
+  }
+  
+  func isDistinct(from other: AccessBase) -> Bool {
+    fatalError()
+  }
+}
+
+/// An `AccessPath` is a pair of a `base: AccessBase` and a `projectionPath: Path`
+/// which denotes the offset of the access from the base in terms of projections.
+struct AccessPath : CustomStringConvertible {
+  var base: AccessBase
+
+  /// address projections only
+  var projectionPath: SmallProjectionPath
+  
+  var description: String {
+    "\(projectionPath): \(base)"
+  }
+  
+  func isDistinct(from: AccessPath) -> Bool {
+    fatalError("TODO")
+  }
+}
+
+/// An `AccessRoot` is the reference (or a value for which the
+/// reference is a stored property) an address originates from.
+struct AccessRoot {
+  var root: Value
+  
+  /// Only valid paths are: `"<sequence of value projections>.<one reference projection>.<sequence of address projections>"`
+  var path: SmallProjectionPath
```

Comment:
Having a path here does not match with the name "AccessRoot".
For me a "root" is just a single Value.
Together with a path, a better name would be "AccessPathFromRoot".